pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,385 @@
+	if !p.isImageAllowed(image) {
+		return credentialprovider.DockerConfig{}
+	}
+
+	p.Lock()
+	defer p.Unlock()
+
+	cachedConfig, found, err := p.getCachedCredentials(image)
+	if err != nil {
+		klog.Errorf("Failed to get cached docker config: %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+
+	if found {
+		return cachedConfig
+	}
+
+	response, err := p.plugin.ExecPlugin(context.Background(), image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+
+	var cacheKey string
+	switch cacheKeyType := response.CacheKeyType; cacheKeyType {
+	case credentials.ImagePluginCacheKeyType:
+		cacheKey = image
+	case credentials.RegistryPluginCacheKeyType:
+		registry := parseRegistry(image)
+		cacheKey = registry
+	case credentials.GlobalPluginCacheKeyType:
+		cacheKey = globalCacheKey
+	default:
+		klog.Errorf("credential provider plugin did not return a valid cacheKeyType: %q", cacheKeyType)
+		return credentialprovider.DockerConfig{}
+	}
+
+	dockerConfig := make(credentialprovider.DockerConfig, len(response.Auth))
+	for matchImage, authConfig := range response.Auth {
+		dockerConfig[matchImage] = credentialprovider.DockerConfigEntry{
+			Username: authConfig.Username,
+			Password: authConfig.Password,
+		}
+	}
+
+	// cache duration was explicitly 0 so don't cache this response at all.
+	if response.CacheDuration != nil && response.CacheDuration.Duration == 0 {
+		return dockerConfig
+	}
+
+	var expiresAt time.Time
+	// nil cache duration means use the default cache duration
+	if response.CacheDuration == nil {
+		if p.defaultCacheDuration == 0 {
+			return dockerConfig
+		}
+
+		expiresAt = time.Now().Add(p.defaultCacheDuration)
+	} else {
+		expiresAt = time.Now().Add(response.CacheDuration.Duration)
+	}
+
+	cachedEntry := &cacheEntry{
+		key:         cacheKey,
+		credentials: dockerConfig,
+		expiresAt:   expiresAt,
+	}
+
+	if err := p.cache.Add(cachedEntry); err != nil {
+		klog.Errorf("Error adding auth entry to cache: %v", err)
+	}
+
+	return dockerConfig
+}
+
+// Enabled always returns true since registration of the plugin via kubelet implies it should be enabled.
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// isImageAllowed returns true if the image matches against the list of allowed matches by the plugin.
+func (p *pluginProvider) isImageAllowed(image string) bool {
+	for _, matchImage := range p.matchImages {
+		if matched, _ := credentialprovider.URLsMatchStr(matchImage, image); matched {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getCachedCredentials returns a credentialprovider.DockerConfig if cached from the plugin.
+func (p *pluginProvider) getCachedCredentials(image string) (credentialprovider.DockerConfig, bool, error) {
+	obj, found, err := p.cache.GetByKey(image)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
```

Comment:
don't we need to check `expiresAt` before using this? (same applies to all the other places we pull a record out of the cache). If so, add a unit test that would have exercised this

Reply:
I assumed that if `expiresAt` passed, the entry wouldn't be in the cache to begin with. Is that not a safe assumption here?

Reply:
sorry, I missed the IsExpired impl above... yeah, this should be fine